### DISCUSSION TITLE
Preset Switching per application

### DISCUSSION THREAD

**Discussion by rimko** [sezanzeb/input-remapper#20] [Ideas]
Hi, i just would like to discuss about preset switching, currently i don't really need it because i can work around with the application that i use but it's quite a popular feature of software of keyboard/mouse manufacture.

In my experience i had experience with two mechanic to switch preset :

- When a specific application is select (in focus)
- When the cursor is hovering on a application.

The second option is my favorite so far, thx to Elecom engineer (Logitech can figure out how to done it i guess ^^). All the solution that i have thy have the same way to deal with preset:

- Main preset apply by default.
- Preset link to a specific application.

**Comment by sezanzeb** [comment]
getting the active window seems to work in X (https://askubuntu.com/a/25658), but not in wayland (https://superuser.com/a/1032426)

**Reply by sezanzeb** [reply]
it seems odd that gnome on wayland is able to know which of the windows is in focus despite numerous stack overflow posts stating that it is not possible. But I only know very little about that stuff

**Reply by rimko** [reply]
Yeah i have try to gather information about the subject but the topic looks quite complicated..
Some other project have the same issue :
[](https://stackoverflow.com/questions/45465016/how-do-i-get-the-active-window-on-gnome-wayland)

**Reply by vanillajonathan** [reply]
The GNOME Shell extension [Dash to Panel](https://github.com/home-sweet-gnome/dash-to-panel) also seem to know.

**Comment by sezanzeb** [comment]
https://github.com/ActivityWatch/activitywatch/issues/92

**Reply by vanillajonathan** [reply]
Interesting post. They seem to have solved it by checking if the session is GNOME then if it is they query gnome-shell and ask it for the currently active/focused window.

**Comment by IggyStanson** [comment]
Hi,
thanks for your amazing tool. I wrote a shell script for automatic preset switching depending on the current active window. So it's basically a "have a different preset per program" function. I have the script on autostart and a different set of keybindings for different programs is insane. 
I am on Manjaro KDE using Xorg. I hardly code any shell, so be kind :-D at least I comment my code
[mouse_tool.zip](https://github.com/sezanzeb/key-mapper/files/6177016/mouse_tool.zip)


Maybe it's helpful,
Thanks!

**Reply by rimko** [reply]
thanks !
Unfortunatly i use fedora with wayland on my workstation.
i hope there will be way to do it on wayland, maybe i will considere switching to a other ditro with xorg when i have to change my workstation.

**Comment by robkam** [comment]
Just to point out that sometimes different browser tabs require different actions. For example in search results I want the mouse wheel to do page up and page down but in maps to zoom in and out.

**Comment by sezanzeb** [comment]
https://github.com/DreadPirateLynx/input-remapper-xautopresets exists for X11

**Reply by DreadPirateLynx** [reply]
Would just like to add that it supports tab-specific  presets, and works in kwin. Now that you’ve made the above GNOME solution known to me, I’ll see if I can incorporate it.
